src/commands.cpp

```diff
@@ -958,67 +957,75 @@ bool CommandGetFile::procresult(Result r)
     }
 }
 
-CommandSetAttr::CommandSetAttr(MegaClient* client, Node* n, SymmCipher* cipher, const char* prevattr)
+CommandSetAttr::CommandSetAttr(MegaClient* client, Node* n, attr_map&& attrMapUpdates, int reqtag)
+    : mAttrMapUpdates(attrMapUpdates)
 {
-    cmd("a");
-    notself(client);
-
-    string at;
-
-    n->attrs.getjson(&at);
-    client->makeattr(cipher, &at, at.c_str(), int(at.size()));
-
-    arg("n", (byte*)&n->nodehandle, MegaClient::NODEHANDLE);
-    arg("at", (byte*)at.c_str(), int(at.size()));
-
+    tag = reqtag;
     h = n->nodehandle;
-    tag = client->reqtag;
-    syncop = prevattr;
+    generationError = API_OK;
 
-    if(prevattr)
-    {
-        pa = prevattr;
-    }
     addToNodePendingCommands(n);
 }
 
-bool CommandSetAttr::procresult(Result r)
+const char* CommandSetAttr::getJSON(MegaClient* client)
 {
-    removeFromNodePendingCommands(h, client);
-#ifdef ENABLE_SYNC
-    if(r.wasError(API_OK) && syncop)
+    // We generate the command just before sending, so it's up to date for any external changes that occured in the meantime
+    // And we can also take into account any changes we have sent for this node that have not yet been applied by actionpackets
+    json.clear();
+    generationError = API_OK;
+
+    cmd("a");
+
+    string at;
+    if (Node* n = client->nodebyhandle(h))
     {
-        Node* node = client->nodebyhandle(h);
-        if(node)
+        AttrMap m = n->attrs;
+
+        // apply these changes for sending, but also any earlier changes that are ahead in the queue
```

Comment:
Yes this one is not just for renames, but other node attributes also.  Eg.  coordinates, custom node attribute as well as name, fingerprint, mtime.  If several change requests for those are called on MegaApi quickly, they may be sent to the servers in a single batch (or spread across two batches).  Each one of those requests to the server will specifiy the entire string of all attributes, encrypted.   So in order for each request not to overwrite the change from the prior request, each subsequent request must also include all prior requested attribute changes as well.   We used to cover this case with Speculative Instant Completion, where we updated the Node's attributes immediately so all changes were included that way when we prepared the next attribute change request.    Now, we don't update the Node's attributes except by actionpacket, so we keep the record of pending attribute changes, and include all changes up to that request in that request - those are calculated just before actually sending, so that we also respect any changes that arrive by actionpacket from other clients.    We need SN tagging to recover from some rare race conditions with other client changing the attributes, but those are not covered currently anyway.

Reply:
Oh, that's a use case I hadn't considered: multiple updates of singles attributes being sent separately. I think it's not a case that will happen very often (perhaps only for `name`), but it's good to have it covered. Thanks for the explanation, your changes make sense and I agree it's more or less the same logic that we used to have.